### PR TITLE
msi for native support

### PR DESCRIPTION
**Pull request by rokmicrosoft**

### CODE REVIEW

**Comment by rokmicrosoft** [comment]
We can't go with [Add-AzureRmAccount -MSI](https://docs.microsoft.com/en-us/powershell/module/azurerm.profile/connect-azurermaccount?view=azurermps-5.5.0) this approach. One reason msiport is variable in our case and second is this feature is only available from 5.5.0. But in our case if vm is MSI enabled then it has backward compatibility as well. So going with token approach.

File: Tasks/AzurePowerShell/Utility.ps1 [hunk 1]
```diff
@@ -30,7 +30,7 @@ function Update-PSModulePathForHostedAgent {
             $hostedAgentAzureModulePath  =  Get-LatestModule -patternToMatch "^azure_[0-9]+\.[0-9]+\.[0-9]+$"   -patternToExtract "[0-9]+\.[0-9]+\.[0-9]+$" -Classic:$true
         }
 
-        if($authScheme -eq 'ServicePrincipal' -or $authScheme -eq '')
+        if($authScheme -eq 'ServicePrincipal' -or $authScheme -eq '' -or $authScheme -eq 'ManagedServiceIdentity')
```
Comment (kmkumaran) [inline Tasks/AzurePowerShell/Utility.ps1]: Would be good to move the empty check to the end.

File: Tasks/Common/VstsAzureHelpers_/InitializeFunctions.ps1 [hunk 2]
```diff
@@ -161,7 +161,32 @@ function Initialize-AzureSubscription {
...
+            $accountId = $env:RELEASE_RELEASEID 
+        }
+        $date = Get-Date -Format o
+        $accountId = -join($accountId, "-", $date)
+        $port = 50342
+        if($Endpoint.Data.MsiPort){
+            $port = $Endpoint.Data.MsiPort
+        }
+        $msiUri =  "http://localhost:$port/oauth2/token"
+        $response = Invoke-WebRequest -Uri $msiUri -Method GET -Body @{resource= $Endpoint.Url} -Headers @{Metadata="true"} -UseBasicParsing
```
Comment (kmkumaran) [inline Tasks/Common/VstsAzureHelpers_/InitializeFunctions.ps1]: All these logic has to change.

File: Tasks/Common/VstsAzureHelpers_/InitializeFunctions.ps1 [hunk 3]
```diff
@@ -161,7 +161,32 @@ function Initialize-AzureSubscription {
...
+        try {
+            Write-Host "##[command]Add-AzureRmAccount  -AccessToken ****** -AccountId $accountId "
+            $null = Add-AzureRmAccount -AccessToken $access_token -AccountId $accountId
+        } catch {
+            # Provide an additional, custom, credentials-related error message.
+            Write-VstsTaskError -Message $_.Exception.Message
+            throw (New-Object System.Exception((Get-VstsLocString -Key AZ_ManagedServiceIdentityError), $_.Exception))
+        }
+        
+        Set-CurrentAzureRMSubscription -SubscriptionId $Endpoint.Data.SubscriptionId -TenantId $Endpoint.Auth.Parameters.TenantId
```
Comment (kmkumaran) [inline Tasks/Common/VstsAzureHelpers_/InitializeFunctions.ps1]: Can you merge with latest changes for TLS done by Rajat.

File: Tasks/AzurePowerShell/Utility.ps1 (hunk 1 above)
Comment (rokmicrosoft) [reply Tasks/AzurePowerShell/Utility.ps1]: Done.

File: Tasks/Common/VstsAzureHelpers_/InitializeFunctions.ps1 [hunk 4]
```diff
@@ -161,11 +161,108 @@ function Initialize-AzureSubscription {
...
+        }
+        else
+        {
+            throw (Get-VstsLocString -Key AZ_MsiAccessNotConfiguredProperlyFailure -ArgumentList $response.StatusCode, $response.StatusDescription)
+        }
+        
+    }
+    catch
+    {
+        $exceptionMessage = $_.Exception.Message.ToString()
```
[the remark this answers is not in the document]
Comment (Ajay-MS) [reply Tasks/Common/VstsAzureHelpers_/InitializeFunctions.ps1]: Sure

[the remark this answers is not in the document]
Comment (rokmicrosoft) [reply Tasks/Common/VstsAzureHelpers_/InitializeFunctions.ps1]: Will take into separate PR.

**Pull request merged**